apiserver/remotefirewaller/ingressaddresswatcher.go

```diff
@@ -197,10 +250,109 @@ func (w *IngressAddressWatcher) mergeAddresses(addresses set.Strings, c params.R
 			}
 		}
 		if !inUse {
-			result.Remove(address)
+			changed = true
 		}
 	}
-	return result, nil
+	return changed, nil
+}
+
+func (w *IngressAddressWatcher) trackUnit(unit Unit) error {
+	machine, err := w.assignedMachine(unit)
+	if errors.IsNotAssigned(err) {
+		logger.Errorf("unit %q entered scope without a machine assigned - addresses will not be tracked", unit)
+		return nil
+	}
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	w.unitToMachine[unit.Name()] = machine.Id()
+	mData, ok := w.machines[machine.Id()]
+	if ok {
+		// We're already watching the machine, just add this unit.
+		mData.units.Add(unit.Name())
+		return nil
+	}
+
+	addressWatcher, err := newMachineAddressWatcher(machine, w.addressChanges)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	w.machines[machine.Id()] = &machineData{
```

Comment:
This doesn't look right - what happens if we have already seen that machine before. This could be the second unit assigned to a particular machine.